### PR TITLE
[DNM] Use secret as Grafana credentials

### LINKED ISSUE #2637: tidb-monitor grafana should set username and password in a Secret
## Feature Request

**Is your feature request related to a problem? Please describe:**

From the example https://github.com/pingcap/tidb-operator/blob/master/manifests/monitor/tidb-monitor.yaml, seems the grafana in tidb-monitor can only set username and password in plain text. This is not good and it is better to use `Secret` to specify these fields.

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

### What problem does this PR solve? 

Fixes #2637 

### Does this PR introduce a user-facing change?:

```release-note
Use secretRef for `Grafana` credentials in `TidbMonitor` instead of plain text
```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 1]
```diff
@@ -118,8 +118,15 @@ type GrafanaSpec struct {
...
-	Username string      `json:"username,omitempty"`
-	Password string      `json:"password,omitempty"`
+
+	// +optional
+	// if username is omitted, Grafana will use `admin` as its username by default.
+	Username *corev1.SecretKeySelector `json:"username,omitempty"`
+
+	// +optional
+	// if password is omitted, Grafana will use `admin` as its password by default.
+	Password *corev1.SecretKeySelector `json:"password,omitempty"`
```
Comment (Yisaer) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:128]: Does it be Forward compatible to the previous TidbMonitor object? What about using a new property to indicate the grafana's password secretRef and make the current Password as Deprecated?

Comment (yeya24) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:128]: Great, I think that makes sense. Will update this week.

**Comment by Yisaer** [comment]
@yeya24  Hi,any progress now?

**Comment by yeya24** [comment]
> @yeya24 Hi,any progress now?

Will work on it today

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 2]
```diff
@@ -118,8 +118,25 @@ type GrafanaSpec struct {
 
 	LogLevel string      `json:"logLevel,omitempty"`
 	Service  ServiceSpec `json:"service,omitempty"`
-	Username string      `json:"username,omitempty"`
-	Password string      `json:"password,omitempty"`
+
+	// +optional
+	// if usernameSecret is omitted, Grafana will use `admin` as its username by default.
+	UsernameSecret *corev1.SecretKeySelector `json:"usernameSecret,omitempty"`
```
Comment (Yisaer) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:124]: I think we can use `Username` and `UsernameSecret` is unnecessary.

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 3]
```diff
@@ -118,8 +118,25 @@ type GrafanaSpec struct {
...
+
+	// +optional
+	// if passwordSecret is omitted, Grafana will use `admin` as its password by default.
+	PasswordSecret *corev1.SecretKeySelector `json:"passwordSecret,omitempty"`
+
+	// +optional
+	// Deprecated
+	// +k8s:openapi-gen=false
+	Username string `json:"username,omitempty"`
```
Comment (Yisaer) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:133]: We don't need to set `Username` as `Deprecated`.

File: pkg/monitor/monitor/util.go [hunk 4]
```diff
@@ -448,32 +433,14 @@ func getMonitorGrafanaContainer(secret *core.Secret, monitor *v1alpha1.TidbMonit
...
-				Name: "GF_SECURITY_ADMIN_PASSWORD",
-				ValueFrom: &core.EnvVarSource{
-					SecretKeyRef: &core.SecretKeySelector{
-						LocalObjectReference: core.LocalObjectReference{
-							Name: secret.Name,
-						},
-						Key: "password",
-					},
-				},
-			},
```
Comment (Yisaer) [inline pkg/monitor/monitor/util.go:472]: I suggest remain these logic for the backward compatibility for 1.1.x user. And add the comment that. We will finally drop these logic in 1.2.0

Comment (yeya24) [reply pkg/monitor/monitor/util.go:472]: I am not very sure. This is a serious security problem, do we really need to prioritize backward compatibility over security in this case?

Comment (Yisaer) [reply pkg/monitor/monitor/util.go:472]: This is a good concern. I think we can break the backward compatibility here. Just mention this the in the request's release note.

Comment (cofyc) [reply pkg/monitor/monitor/util.go:472]: Especially in patch versions, we should not introduce break changes. The security problem is not unavoidable, we can encourage people to use the secret to store password after this.

We can remove `password` field in next minor version v1.12.

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 5]
```diff
@@ -121,8 +121,18 @@ type GrafanaSpec struct {
...
-	Password string      `json:"password,omitempty"`
+
+	// +optional
+	// if passwordSecret is omitted, Grafana will use `admin` as its password by default.
+	PasswordSecret *corev1.SecretKeySelector `json:"passwordSecret,omitempty"`
+
+	Username string `json:"username,omitempty"`
+
+	// +optional
+	// Deprecated
```
Comment (cofyc) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:132]:
```suggestion
	// Deprecated in v1.1.2 for security concerns, planned for removal in v1.2. Use `passwordSecret` instead.
```

File: pkg/monitor/monitor/util.go (hunk 4 above)
Comment (DanielZhangQD) [reply pkg/monitor/monitor/util.go:472]: Agree with @cofyc , we should maintain the backward compatibility here, and we can make it clear in the doc & comments that the secret should be used.

**Comment by Yisaer** [comment]
@yeya24 Hi, after discussion with other maintainers, we think it's better not to introduce break changes in the minor version. We could remove the password in the next major version.

**Comment by yeya24** [comment]
> @yeya24 Hi, after discussion with other maintainers, we think it's better not to introduce break changes in the minor version. We could remove the password in the next major version.

Understand. Sorry for the late response last week. I will update it today.

**Comment by DanielZhangQD** [comment]
> > @yeya24 Hi, after discussion with other maintainers, we think it's better not to introduce break changes in the minor version. We could remove the password in the next major version.
> 
> Understand. Sorry for the late response last week. I will update it today.

@yeya24 Just a friendly ping.

**Comment by yeya24** [comment]
> > > @yeya24 Hi, after discussion with other maintainers, we think it's better not to introduce break changes in the minor version. We could remove the password in the next major version.
> > 
> > 
> > Understand. Sorry for the late response last week. I will update it today.
> 
> @yeya24 Just a friendly ping.

Got it, will try to update it this week.

**Comment by DanielZhangQD** [comment]
@yeya24 Are you still working on this?

**Comment by yeya24** [comment]
Sorry I don't have time to work on this recently. This is stale and would be good if someone else can pick this up.

**Comment by DanielZhangQD** [comment]
> Sorry I don't have time to work on this recently. This is stale and would be good if someone else can pick this up.

OK. Thanks!

**Comment by DanielZhangQD** [comment]
Close this temporarily, we may have someone to continue the work later.

**Pull request closed without merging**